crates/rattler_installs_packages/src/index/package_database.rs

```diff
@@ -111,6 +127,390 @@ impl PackageDb {
+        let normalized_package_name = p.into();
+
+        let parsed_url = ParsedUrl::new(&url)?;
+
+        let git_source = GitSource {
+            url: parsed_url.git_url,
+            rev: parsed_url.revision,
+        };
+
+        let temp_dir = tempdir().unwrap();
+
+        let mut location = git_clone(&git_source, &temp_dir).into_diagnostic()?;
+
+        if let Some(subdirectory) = parsed_url.subdirectory {
+            location.push(&subdirectory);
+            if !location.exists() {
+                return Err(miette::miette!(
+                    "Requested subdirectory fragment {:?} can't be located at following url {:?}",
+                    subdirectory,
+                    url
+                ));
+            }
+        };
+
+        let (wheel_metadata, filename) = self
+            .get_stree_from_file_path(
+                &normalized_package_name,
+                url.clone(),
+                Some(location),
+                wheel_builder,
+            )
+            .await?;
+
+        let requires_python = wheel_metadata.1.requires_python;
+
+        let dist_info_metadata = DistInfoMetadata {
+            available: false,
+            hashes: ArtifactHashes::default(),
+        };
+
+        let yanked = Yanked {
+            yanked: false,
+            reason: None,
+        };
+
+        let project_hash = ArtifactHashes {
+            sha256: Some(compute_bytes_digest::<Sha256>(url.as_str().as_bytes())),
+        };
+
+        let artifact_info = ArtifactInfo {
+            filename,
+            url: url.clone(),
+            hashes: Some(project_hash),
+            requires_python,
+            dist_info_metadata,
+            yanked,
+        };
+
+        let mut result: IndexMap<PypiVersion, Vec<ArtifactInfo>> = Default::default();
+        result
+            .entry(PypiVersion::Url(url))
+            .or_default()
+            .push(artifact_info.clone());
+
+        self.put_metadata_in_cache(&artifact_info, &wheel_metadata.0)
+            .await?;
+
+        Ok(self
+            .artifacts
+            .insert(normalized_package_name.clone(), Box::new(result)))
+    }
+
+    /// Get artifact directly from file, vcs, or url
+    pub async fn get_artifact_by_url<'a, 'i, P: Into<NormalizedPackageName>>(
+        &self,
+        p: P,
+        url: Url,
+        wheel_builder: &WheelBuilder<'a, 'i>,
+    ) -> miette::Result<&IndexMap<PypiVersion, Vec<ArtifactInfo>>> {
+        //conver to str
+        let p = p.into();
+
+        if let Some(cached) = self.artifacts.get(&p) {
+            Ok(cached)
+        } else if url.scheme() == "file" {
+            self.get_file_artifact(p, url, wheel_builder).await
+        } else if url.scheme() == "https" {
+            self.get_http_artifact(p, url, wheel_builder).await
+        } else if url.scheme() == "git+https" || url.scheme() == "git+file" {
+            self.get_git_artifact(p, url, wheel_builder).await
```

Comment:
Trying to think of a name to make these a bit more connected, because pypi itself is also a Url of course. So maybe something like `get_direct_url_file_artifact`, `get_direct_url_http` something like that?